### ISSUE TITLE
Image upload: Show progress feedback

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#2684]
When uploading an image, the upload might take several seconds or more, particularly if the image is large or the connection is slow.

We should give some visual feedback, like a progress bar, to indicate that the upload is happening.

**Comment by agzuniverse** [comment]
I'll attempt to solve this, but not claiming since I'm not sure if I will get a solution up fast enough.

**Comment by borisyankov** [comment]
Showing a progress would be quite involved (and ideally a cancel 'X' button)
But we can improve upon the current state with a much smaller feature.
We can initially just show placeholder with an indefinite spinner inside.

Showing the actual progress involves this:
* using some other lib to upload, since we are not getting any progress feedback from the basic `fetch` API
* a way to track all active uploads
* a custom way to send data about progress about these uploads to the web view

**Comment by borisyankov** [comment]
The message prerender happens in 
https://github.com/kunall17/zulip-markdown-parser
which might also require changes.

**Comment by gnprice** [comment]
Copying back here a comment I made last year on #3118:

I think an image-shaped and -sized placeholder with a spinner -- Zulip-green on neutral background, like most of our spinners -- would probably work well.

Experimenting with it now, we do have a spinner: the usual outbox spinner shows up there just fine. So I'm actually a little curious just how users respond to the current UI -- is that spinner just too subtle? (It is pretty small.) Or is there another issue?

Hmm, in fact: what we *don't* have a spinner for is *downloading*, and rendering, the image. Experimenting with this on [the #3118 branch], I still have the experience that it feels a bit stuck and I briefly wonder if it's making progress, because that part is slow and the message just looks blank. (In fact the green box [which the PR added while uploading] might barely make it on screen for a frame or two, then it's blank for hundreds of ms and then fills in gradually from the top over a second or so.)

I think one thing that would improve the situation is if we can get the image to show as a gray box or something when it isn't downloaded yet -- then at least it would have clearly-defined boundaries in its own shape, so it's clear there is supposed to be an image there and things aren't just weirdly all blank.

**Comment by gnprice** [comment]
@timabbott ran into this issue today, and reported [in chat](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/photo.20rotation/near/787915):

> Because the file was large (and thus took some time to upload) and we apparently don't local echo or provide any other feedback, it looked like the upload didn't work at all, which confused me into uploading duplicate copies. We should really do something, whether it be an upload progress bar (might be the best solution), loading spinner, local echo, or something. (Though my guess is we're not local echoing because we only send the message once it's been created, and that only happens when the upload finishes)

So if it is still the case (as I reported in the previous comment) that we show the outbox spinner while uploading, then it's also still the case that that's quite subtle and easy to miss.

The message consists entirely of a link, like `[foo.jpg](/user_uploads/1/a/asdfqwer/foo.jpg)`. On the web these look like the link text, followed by the image... but I believe we render them as just the image. So one point is that the outbox message (as local-echoed) is probably completely blank before the image is uploaded... there's nothing there *but* the spinner. That likely contributes to making it feel like there's no feedback.

**Comment by gnprice** [comment]
I just tested uploading a moderately-big photo (2.5MB), and then a video (24.5MB), for investigating #3624. The video took a little time (at least 5s), which gave me a chance to get a good look at what the message list looks like while we're doing the upload.

There was no spinner during that time. It wasn't a scrolling issue, either -- I had time to try several times to scroll down (in case an outbox-message with spinner had appeared below the viewport), and there was nothing there. The message list looked exactly the same way it had before I went into the file-upload UI. So that means this issue is still fully open as reported.

A spinner did briefly appear for a fraction of a second, at the end just before the message appeared fully without a spinner. I believe that represents the time when we'd completed the upload, and went on to send the actual message.

In case it matters for the repro: this was all while using the file-upload flow, with the "document" icon in the compose menu. But I expect the camera and gallery flows would behave the same way as this one.

**Comment by chrisbobbe** [comment]
@gnprice [said](https://github.com/zulip/zulip-mobile/issues/2684#issuecomment-534218632):
> I think an image-shaped and -sized placeholder with a spinner -- Zulip-green on neutral background, like most of our spinners -- would probably work well.

Agreed. We'll need to use the image's dimensions to size the the placeholder; this should be fine.

**Comment by chrisbobbe** [comment]
One other possible direction is to do it the web app's way — nothing actually happens in the message list. The actual text in the compose box is replaced with something like `[Uploading fileName.png...]` during the upload, and a progress indicator (could just be a static spinner to start out with) appears above that.

**Comment by chrisbobbe** [comment]
This will be substantially helped by #4590.

**Comment by chrisbobbe** [comment]
I've just closed #4129 as likely to be superseded by #4590. If it eventually turns out that #4129 was the better direction to go, we should consider reopening it. 🙂

**Comment by devZer0** [comment]
yes please.  i see many users accidentally uploading the same image multiple times because they think their upload did not work. before i knew this is by design, i did this, too.    some indication of upload progress would be very nice.

**Comment by chrisbobbe** [comment]
We've had another report of this: https://chat.zulip.org/#narrow/stream/48-mobile/topic/Bug.20reports/near/1292548
